src/odemis/acq/acqmng.py

```diff
@@ -83,6 +85,202 @@ def acquire(streams, settings_obs=None):
+        """
+        remaining_t = self.estimate_total_duration()
+        acquired_data = []
+        # iterate through streams
+        for stream in self._streams:
+            zstack = []
+            if stream not in self._zlevels: 
+                try:
+                    # acquire this single stream, and get the data
+                    self._single_acqui_f = acquire([stream], self._settings_obs)
+                    data, exp = self._single_acqui_f.result()
+
+                    # check if cancellation happened while the acquiring future is working
+                    if self._future_state == CANCELLED:
+                        raise CancelledError()
+
+                    # check on the acquired data
+                    if not data:
+                        logging.warning("The acquired data array for stream %s is empty", stream)
+                    else:
+                        logging.info("The acquisition for stream %s is done", stream)
+
+                except CancelledError:
+                    raise 
+                except Exception as e:
+                    logging.exception("The acquisition failed because %s" %e)
+                    # return what was acquired so far  
+                    return acquired_data, e
+              
+                acquired_data.append(data[0])
+                # update the remaining time
+                remaining_t -= stream.estimateAcquisitionTime()
+                self._main_future.set_end_time(time.time() + remaining_t)
+
+            else: 
+                # for each stream, iterate through zlevels
+                for i, z in enumerate(self._zlevels[stream]):
+                    # move the focuser
+                    self._actuator_f = stream.focuser.moveAbs({"z": z})
+                    self._actuator_f.result()
+
+                    # check if cancellation happened while the actuator future is working
+                    if self._future_state == CANCELLED:
+                        raise CancelledError()
+
+                    # subtract one zstep time
+                    if i != len(self._zlevels[stream]) - 1:
+                        remaining_t -= self._zstep_duration[stream]
+                        self._main_future.set_end_time(time.time() + remaining_t)
+
+                    try:
+                        # acquire this single stream, and get the data
+                        self._single_acqui_f = acquire([stream], self._settings_obs)
+                        data, exp = self._single_acqui_f.result()
+
+                        # check if cancellation happened while the acquiring future is working
+                        if self._future_state == CANCELLED:
+                            raise CancelledError()
+
+                        # check on the acquired data
+                        if not data:
+                            logging.warning("The acquired data array for stream %s is empty", stream)
```

Comment:
You show a warning here, but in such case the line `zstack.append(data[0])` will completely fail. => don't make that line fail.